sdk/core/Azure.Core.TestFramework/src/ProxyTransport.cs

```diff
@@ -45,23 +45,49 @@ public ProxyTransport(TestProxy proxy, HttpPipelineTransport transport, TestReco
             _filter = filter;
         }
 
-        public override void Process(HttpMessage message)
-        {
-            RedirectToTestProxy(message);
-            _innerTransport.Process(message);
-            ProcessResponseAsync(message, false).EnsureCompleted();
-        }
+        public override void Process(HttpMessage message) =>
+            ProcessAsyncInternalAsync(message, false).EnsureCompleted();
 
-        public override async ValueTask ProcessAsync(HttpMessage message)
+        public override async ValueTask ProcessAsync(HttpMessage message) =>
+            await ProcessAsyncInternalAsync(message, true);
+
+        private async Task ProcessAsyncInternalAsync(HttpMessage message, bool async)
         {
-            RedirectToTestProxy(message);
-            await _innerTransport.ProcessAsync(message);
-            await ProcessResponseAsync(message, true);
+            try
+            {
+                RedirectToTestProxy(message);
+                if (async)
+                {
+                    await _innerTransport.ProcessAsync(message);
+                }
+                else
+                {
+                    _innerTransport.Process(message);
+                }
+
+                await ProcessResponseAsync(message, true);
+            }
+            finally
+            {
+                // revert the original URI - this is important for tests that rely on aspects of the URI in the pipeline
+                // e.g. KeyVault caches tokens based on URI
+                message.Request.Headers.TryGetValue("x-recording-upstream-base-uri", out string original);
+
+                var originalBaseUri = new Uri(original);
+                message.Request.Uri.Scheme = originalBaseUri.Scheme;
+                message.Request.Uri.Host = originalBaseUri.Host;
+                message.Request.Uri.Port = originalBaseUri.Port;
```

Comment:
Would it be any better to do this with UriBulder?

Reply:
the Uri on the request is a RequestUriBuilder already.